rcl_action/src/rcl_action/action_server.c

```diff
@@ -0,0 +1,759 @@
+  return goal_handles[num_goal_handles];
+}
+
+rcl_ret_t
+rcl_action_publish_feedback(
+  const rcl_action_server_t * action_server,
+  void * ros_feedback)
+{
+  if (!rcl_action_server_is_valid(action_server)) {
+    return RCL_RET_ACTION_SERVER_INVALID;  // error already set
+  }
+  RCL_CHECK_ARGUMENT_FOR_NULL(ros_feedback, RCL_RET_INVALID_ARGUMENT);
+
+  rcl_ret_t ret = rcl_publish(&action_server->impl->feedback_publisher, ros_feedback);
+  if (RCL_RET_OK != ret) {
+    return RCL_RET_ERROR;  // error already set
+  }
+  return RCL_RET_OK;
+}
+
+rcl_ret_t
+rcl_action_get_goal_status_array(
+  const rcl_action_server_t * action_server,
+  rcl_action_goal_status_array_t * status_message)
+{
+  if (!rcl_action_server_is_valid(action_server)) {
+    return RCL_RET_ACTION_SERVER_INVALID;  // error already set
+  }
+  RCL_CHECK_ARGUMENT_FOR_NULL(status_message, RCL_RET_INVALID_ARGUMENT);
+
+  // If number of goals is zero, then we don't need to do any allocation
+  size_t num_goals = action_server->impl->num_goal_handles;
+  if (0 == num_goals) {
+    status_message->msg.status_list.size = 0u;
+    return RCL_RET_OK;
+  }
+
+  // Allocate status array
+  rcl_allocator_t allocator = action_server->impl->options.allocator;
+  rcl_ret_t ret = rcl_action_goal_status_array_init(status_message, num_goals, allocator);
+  if (RCL_RET_OK != ret) {
+    if (RCL_RET_BAD_ALLOC == ret) {
+      return RCL_RET_BAD_ALLOC;
+    }
+    return RCL_RET_ERROR;
+  }
+
+  // Populate status array
+  for (size_t i = 0; i < num_goals; ++i) {
+    ret = rcl_action_goal_handle_get_info(
+      action_server->impl->goal_handles[i], &status_message->msg.status_list.data[i].goal_info);
+    if (RCL_RET_OK != ret) {
+      return RCL_RET_ERROR;
+    }
+    ret = rcl_action_goal_handle_get_status(
+      action_server->impl->goal_handles[i], &status_message->msg.status_list.data[i].status);
+    if (RCL_RET_OK != ret) {
+      return RCL_RET_ERROR;
+    }
+  }
+  return RCL_RET_OK;
+}
+
+rcl_ret_t
+rcl_action_publish_status(
+  const rcl_action_server_t * action_server,
+  const void * status_message)
+{
+  if (!rcl_action_server_is_valid(action_server)) {
+    return RCL_RET_ACTION_SERVER_INVALID;  // error already set
+  }
+  RCL_CHECK_ARGUMENT_FOR_NULL(status_message, RCL_RET_INVALID_ARGUMENT);
+
+  rcl_ret_t ret = rcl_publish(&action_server->impl->status_publisher, status_message);
+  if (RCL_RET_OK != ret) {
+    return RCL_RET_ERROR;  // error already set
+  }
+  return RCL_RET_OK;
+}
+
+rcl_ret_t
+rcl_action_take_result_request(
+  const rcl_action_server_t * action_server,
+  void * ros_result_request)
+{
+  TAKE_SERVICE_REQUEST(result);
+}
+
+rcl_ret_t
+rcl_action_send_result_response(
+  const rcl_action_server_t * action_server,
+  void * ros_result_response)
+{
+  SEND_SERVICE_RESPONSE(result);
+}
+
+rcl_ret_t
+rcl_action_clear_expired_goals(
+  const rcl_action_server_t * action_server,
+  size_t * num_expired)
```

Comment:
My original thinking was that since the action server is the entity creating the goal handles, it should be the one responsible for destroying them.

In a rcl goal handle is finalized, then the goal handle becomes invalid (ie. `rcl_action_goal_handle_is_valid()` returns false) and so, I'm pretty sure, any rcl operations involving the goal handle should fail with an error code. So at least, the C++ wrapper of a goal handle wouldn't be using a free'd goal handle without getting an explicit error.

Is it enough for the client libraries to check their goal handles with `rcl_action_goal_handle_is_valid()` and/or get a fresh list of goal handles with `rcl_action_server_get_goal_handles()` ?

Reply:
>Is it enough for the client libraries to check their goal handles with `rcl_action_goal_handle_is_valid()`

I guess the question is what does a C++ goal handle wrapper have a reference to? If it has an `rcl_action_goal_handle_t * rcl_handle_;` pointing to storage owned by [`rcl_action_server_impl_t`](https://github.com/ros2/rcl/pull/323/files#diff-d6a1cca7ac458b893d4cd7204485edf2R47), then `rcl_action_goal_handle_is_valid(rcl_handle_)` isn't safe because the memory it's pointing to could change any time `impl->goal_handles` is resized. It seems like it needs to be the other way around where the `C++` goal handle wrapper has an `rcl_action_goal_handle_t` and `rcl_action_server_t` is the one with a pointer to it.

Reply:
I see; makes sense now.

Reply:
Hmm, so just to be clear, rcl* will handle all incoming result requests by first checking for goal handle validity, rejecting the request once the goal handle is no longer valid and only then deallocating it, right? I'd be somewhat more comfortable with just marking the goal handle as expired to keep finalization and deallocation close together, but this will do to.

Reply:
@hidmic I was going to refactor rcl action server so that goal handle finalization does not happen here. It will be the responsibility of the caller to finalize and deallocate goal handles. rcl will only resize the internal array as goals become expired.

Reply:
Hmm, then how does the calling library knows when a goal handle has expired?

Reply:
> Hmm, then how does the calling library know when a goal handle has expired?
 
What about if we return a list of goal handles as part of the expire function.

Reply:
Sounds like an option. Though judging by @sloretz rclcpp API skeleton, I'd imagine that having the client library wrappers checking if the goal handle they own has expired or not may be simpler. But we can somehow reverse the mapping to know which wrappers have to go away when goal handles expire.

Reply:
The client library can call `rcl_action_server_goal_exists()` to check if a previously accepted goal has expired.

Reply:
Removed finalize of expired goal handles in 1140112.
It is expected that the client library will finalize goal handles either after they expire (which can be checked with `rcl_action_server_goal_exists()`) or after the action server is finalized.

Note, the rcl goal handles are still initialized in `rcl_action_accept_new_goal()`. I was going to update the API to take an already initialized goal handle as an argument, but then we lose the time stamp created when the goal was accepted:

https://github.com/ros2/rcl/blob/11401120af69d5652ffb7a2ad46502159941039b/rcl_action/src/rcl_action/action_server.c#L362-L370

If we want the client library to pass the goal handle to rcl, then we can

1. Create a setter function for the goal handle's info and override the info in `rcl_action_accept_goal_info`, or
2. Leave the responsibility to the client library to use the same clock and stamp the goal handles properly.